File: src/core/lib/scripting/apis/__tests__/EventAPI.test.ts
/**
 * EventAPI Tests
 */

import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { createEventAPI } from '../EventAPI';
import { emitter } from '@/core/lib/events';

describe('EventAPI', () => {
  const entityId = 1;
  let eventAPI: ReturnType<typeof createEventAPI>;

  beforeEach(() => {
    eventAPI = createEventAPI(entityId);
  });

  afterEach(() => {
    // Clean up all event listeners
    emitter.all.clear();
  });

  it('should create an event API instance', () => {
    expect(eventAPI).toBeDefined();
    expect(eventAPI.on).toBeInstanceOf(Function);
    expect(eventAPI.off).toBeInstanceOf(Function);
    expect(eventAPI.emit).toBeInstanceOf(Function);
  });

  it('should subscribe to events', () => {
    const handler = vi.fn();
    const unsubscribe = eventAPI.on('test:event', handler);

    expect(unsubscribe).toBeInstanceOf(Function);

    // Emit event
    eventAPI.emit('test:event', { data: 'test' });

    expect(handler).toHaveBeenCalledOnce();
    expect(handler).toHaveBeenCalledWith({ data: 'test' });
  });

  it('should unsubscribe from events', () => {
    const handler = vi.fn();
    const unsubscribe = eventAPI.on('test:event', handler);

    // Emit event - should be called
    eventAPI.emit('test:event', { data: 'test1' });
    expect(handler).toHaveBeenCalledOnce();

    // Unsubscribe
    unsubscribe();

    // Emit event again - should not be called
    eventAPI.emit('test:event', { data: 'test2' });
    expect(handler).toHaveBeenCalledOnce(); // Still only once
  });

  it('should unsubscribe using off method', () => {
    const handler = vi.fn();
    eventAPI.on('test:event', handler);

    // Emit event - should be called
    eventAPI.emit('test:event', { data: 'test1' });
    expect(handler).toHaveBeenCalledOnce();

    // Unsubscribe using off
    eventAPI.off('test:event', handler);

    // Emit event again - should not be called
    eventAPI.emit('test:event', { data: 'test2' });
    expect(handler).toHaveBeenCalledOnce(); // Still only once
  });

  it('should handle multiple subscribers', () => {
    const handler1 = vi.fn();
    const handler2 = vi.fn();

    eventAPI.on('test:event', handler1);
    eventAPI.on('test:event', handler2);

    eventAPI.emit('test:event', { data: 'test' });

    expect(handler1).toHaveBeenCalledOnce();
    expect(handler2).toHaveBeenCalledOnce();
  });

  it('should handle multiple event types', () => {
    const handler1 = vi.fn();
    const handler2 = vi.fn();

    eventAPI.on('event:one', handler1);
    eventAPI.on('event:two', handler2);

    eventAPI.emit('event:one', { data: '1' });
    expect(handler1).toHaveBeenCalledOnce();
    expect(handler2).not.toHaveBeenCalled();

    eventAPI.emit('event:two', { data: '2' });
    expect(handler1).toHaveBeenCalledOnce();
    expect(handler2).toHaveBeenCalledOnce();
  });

  it('should emit events with different payload types', () => {
    const handler = vi.fn();
    eventAPI.on('test:event', handler);

    // String payload
    eventAPI.emit('test:event', 'string');
    expect(handler).toHaveBeenCalledWith('string');

    // Number payload
    eventAPI.emit('test:event', 42);
    expect(handler).toHaveBeenCalledWith(42);

    // Object payload
    eventAPI.emit('test:event', { key: 'value' });
    expect(handler).toHaveBeenCalledWith({ key: 'value' });

    // Array payload
    eventAPI.emit('test:event', [1, 2, 3]);
    expect(handler).toHaveBeenCalledWith([1, 2, 3]);
  });
});

File: rust/engine/crates/scripting/src/apis/event_api_test.rs
#[cfg(test)]
mod tests {
    use super::super::event_api::*;
    use mlua::Lua;

    #[test]
    fn test_event_api_registration() {
        clear_all_events();
        let lua = Lua::new();
        let entity_id = 1;

        // Should register without errors
        assert!(register_event_api(&lua, entity_id).is_ok());

        // Should have events global
        let result: Result<mlua::Table, _> = lua.globals().get("events");
        assert!(result.is_ok());

        let events = result.unwrap();
        assert!(events.contains_key("on").unwrap());
        assert!(events.contains_key("off").unwrap());
        assert!(events.contains_key("emit").unwrap());
    }

    #[test]
    fn test_event_on_registers_handler() {
        clear_all_events();
        let lua = Lua::new();
        let entity_id = 1;

        register_event_api(&lua, entity_id).unwrap();

        // Register a handler
        let result: Result<u64, _> = lua
            .load(
                r#"
            local handlerId = events:on("test:event", function(data)
                -- Handler function
            end)
            return handlerId
        "#,
            )
            .eval();

        assert!(result.is_ok());
        let handler_id = result.unwrap();
        assert!(handler_id > 0, "Handler ID should be positive");
    }

    #[test]
    fn test_event_emit_calls_handler() {
        clear_all_events();
        let lua = Lua::new();
        let entity_id = 1;

        register_event_api(&lua, entity_id).unwrap();

        // Set up handler that sets a global variable
        lua.load(
            r#"
            callCount = 0
            receivedData = nil

            events:on("test:event", function(data)
                callCount = callCount + 1
                receivedData = data
            end)
        "#,
        )
        .exec()
        .unwrap();

        // Emit the event
        lua.load(
            r#"
            events:emit("test:event", { value = 42 })
        "#,
        )
        .exec()
        .unwrap();

        // Check handler was called
        let call_count: i32 = lua.globals().get("callCount").unwrap();
        assert_eq!(call_count, 1, "Handler should be called once");

        // Check data was received
        let received: mlua::Table = lua.globals().get("receivedData").unwrap();
        let value: i32 = received.get("value").unwrap();
        assert_eq!(value, 42, "Handler should receive correct data");
    }

    #[test]
    fn test_event_multiple_handlers() {
        clear_all_events();
        let lua = Lua::new();
        let entity_id = 1;

        register_event_api(&lua, entity_id).unwrap();

        // Register multiple handlers
        lua.load(
            r#"
            callCount1 = 0
            callCount2 = 0

            events:on("test:event", function(data)
                callCount1 = callCount1 + 1
            end)

            events:on("test:event", function(data)
                callCount2 = callCount2 + 1
            end)
        "#,
        )
        .exec()
        .unwrap();

        // Emit the event
        lua.load(
            r#"
            events:emit("test:event", { value = 100 })
        "#,
        )
        .exec()
        .unwrap();

        // Both handlers should be called
        let call_count1: i32 = lua.globals().get("callCount1").unwrap();
        let call_count2: i32 = lua.globals().get("callCount2").unwrap();
        assert_eq!(call_count1, 1, "First handler should be called");
        assert_eq!(call_count2, 1, "Second handler should be called");
    }

    #[test]
    fn test_event_off_unregisters_handler() {
        clear_all_events();
        let lua = Lua::new();
        let entity_id = 1;

        register_event_api(&lua, entity_id).unwrap();

        // Register handler and get its ID
        lua.load(
            r#"
            callCount = 0

            handlerId = events:on("test:event", function(data)
                callCount = callCount + 1
            end)
        "#,
        )
        .exec()
        .unwrap();

        // Emit once
        lua.load(
            r#"
            events:emit("test:event", {})
        "#,
        )
        .exec()
        .unwrap();

        let call_count_1: i32 = lua.globals().get("callCount").unwrap();
        assert_eq!(call_count_1, 1);

        // Unregister
        lua.load(
            r#"
            events:off("test:event", handlerId)
        "#,
        )
        .exec()
        .unwrap();

        // Emit again - should not increase count
        lua.load(
            r#"
            events:emit("test:event", {})
        "#,
        )
        .exec()
        .unwrap();

        let call_count_2: i32 = lua.globals().get("callCount").unwrap();
        assert_eq!(call_count_2, 1, "Handler should not be called after off()");
    }

    // TODO: This test fails due to global state pollution between tests
    // Need to redesign how event handlers are tracked per-test to fix this
    // #[test]
    // fn test_event_different_event_types() {
    //     clear_all_events();
    //     let lua = Lua::new();
    //     let entity_id = 1;
    //
    //     register_event_api(&lua, entity_id).unwrap();
    //
    //     // Register handlers for different events
    //     lua.load(r#"
    //         event1Count = 0
    //         event2Count = 0
    //
    //         events:on("event:type1", function(data)
    //             event1Count = event1Count + 1
    //         end)
    //
    //         events:on("event:type2", function(data)
    //             event2Count = event2Count + 1
    //         end)
    //     "#).exec().unwrap();
    //
    //     // Emit type1
    //     lua.load(r#"
    //         events:emit("event:type1", {})
    //     "#).exec().unwrap();
    //
    //     let count1: i32 = lua.globals().get("event1Count").unwrap();
    //     let count2: i32 = lua.globals().get("event2Count").unwrap();
    //     assert_eq!(count1, 1, "Type1 handler should be called");
    //     assert_eq!(count2, 0, "Type2 handler should not be called");
    //
    //     // Emit type2
    //     lua.load(r#"
    //         events:emit("event:type2", {})
    //     "#).exec().unwrap();
    //
    //     let count1: i32 = lua.globals().get("event1Count").unwrap();
    //     let count2: i32 = lua.globals().get("event2Count").unwrap();
    //     assert_eq!(count1, 1, "Type1 handler should still be 1");
    //     assert_eq!(count2, 1, "Type2 handler should now be called");
    // }

    #[test]
    fn test_event_no_handlers() {
        clear_all_events();
        let lua = Lua::new();
        let entity_id = 1;

        register_event_api(&lua, entity_id).unwrap();

        // Emit to event with no handlers - should not error
        let result = lua
            .load(
                r#"
            events:emit("nonexistent:event", { data = "test" })
        "#,
            )
            .exec();

        assert!(
            result.is_ok(),
            "Emitting to non-existent event should not error"
        );
    }

    #[test]
    fn test_event_payload_types() {
        clear_all_events();
        let lua = Lua::new();
        let entity_id = 1;

        register_event_api(&lua, entity_id).unwrap();

        // Test with different payload types
        lua.load(
            r#"
            stringPayload = nil
            numberPayload = nil
            tablePayload = nil
            boolPayload = nil

            events:on("test:string", function(data) stringPayload = data end)
            events:on("test:number", function(data) numberPayload = data end)
            events:on("test:table", function(data) tablePayload = data end)
            events:on("test:bool", function(data) boolPayload = data end)

            events:emit("test:string", "hello")
            events:emit("test:number", 42)
            events:emit("test:table", { x = 1, y = 2 })
            events:emit("test:bool", true)
        "#,
        )
        .exec()
        .unwrap();

        // Verify payloads
        let string_val: String = lua.globals().get("stringPayload").unwrap();
        assert_eq!(string_val, "hello");

        let number_val: i32 = lua.globals().get("numberPayload").unwrap();
        assert_eq!(number_val, 42);

        let table_val: mlua::Table = lua.globals().get("tablePayload").unwrap();
        let x: i32 = table_val.get("x").unwrap();
        let y: i32 = table_val.get("y").unwrap();
        assert_eq!(x, 1);
        assert_eq!(y, 2);

        let bool_val: bool = lua.globals().get("boolPayload").unwrap();
        assert_eq!(bool_val, true);
    }

    #[test]
    fn test_cleanup_event_api() {
        clear_all_events();
        let lua = Lua::new();
        let entity_id = 1;

        register_event_api(&lua, entity_id).unwrap();

        // Register handlers
        lua.load(
            r#"
            events:on("test:event", function(data) end)
            events:on("test:event2", function(data) end)
        "#,
        )
        .exec()
        .unwrap();

        // Cleanup
        cleanup_event_api(&lua, entity_id).unwrap();

        // Handlers should be removed from registry
        // (This is tested implicitly - if cleanup fails, Lua would hold references)

        // After cleanup, should be safe to register again
        let result = register_event_api(&lua, entity_id);
        assert!(result.is_ok());
    }

    // TODO: Cross-VM events require a different architecture
    // Lua functions can't be called across different Lua VMs due to mlua's safety model
    // Need to implement message-passing or use a shared Lua VM for all entities
    // #[test]
    // fn test_cross_entity_events() {
    //     clear_all_events();
    //
    //     // Each entity gets its own Lua VM (realistic scenario)
    //     let lua1 = Lua::new();
    //     let lua2 = Lua::new();
    //
    //     // Register API for two different entities (different Lua VMs)
    //     register_event_api(&lua1, 1).unwrap();
    //     register_event_api(&lua2, 2).unwrap();
    //
    //     // Entity 1 listens
    //     lua1.load(r#"
    //         entity1Received = nil
    //         events:on("cross:entity", function(data)
    //             entity1Received = data
    //         end)
    //     "#).exec().unwrap();
    //
    //     // Entity 2 emits (from its own Lua VM)
    //     lua2.load(r#"
    //         events:emit("cross:entity", { sender = "entity2" })
    //     "#).exec().unwrap();
    //
    //     // Entity 1 should receive (handlers are called via global EVENT_BUS)
    //     let received: mlua::Table = lua1.globals().get("entity1Received").unwrap();
    //     let sender: String = received.get("sender").unwrap();
    //     assert_eq!(sender, "entity2", "Cross-entity communication should work");
    // }
}
